apps/wallet/src/engine/runners/solana/history.ts

```diff
@@ -0,0 +1,318 @@
+	const id = signature;
+	const transactionType = getTransactionType(
+		transaction.message.accountKeys[0].pubkey,
+		walletAddress,
+	);
+	const { sender, receiver, amount, token } =
+		(await getTransactionBalances(connection, parsedTransaction)) || {};
+
+	const tokenForFee = await getTransactionFee(walletAddress);
+	const fee = (meta?.fee || 0) / 10 ** (tokenForFee?.account.decimals || 9);
+	const date = blockTime ? new Date(blockTime * 1000) : new Date();
+
+	return {
+		_id,
+		type: 'History',
+		id,
+		signature,
+		network: Networks.solana,
+		transactionType,
+		status: 'Success',
+		sender,
+		receiver,
+		amount,
+		token,
+		tokenForFee,
+		fee,
+		date,
+	} as TransactionHistoryDocument;
+};
+
+const getTransactionType = (sender: PublicKey, walletAddress: string) => {
+	return sender.toString() === walletAddress ? 'Sent' : 'Received';
+};
+
+const getTransactionBalances = async (
+	connection: Connection,
+	parsedTransaction: ParsedTransactionWithMeta,
+) => {
+	const { meta, transaction } = parsedTransaction;
+	if (!meta) return;
+	if (!meta.preTokenBalances || !meta.postTokenBalances) return;
+
+	const isSplTokenTransaction =
+		meta.preTokenBalances?.length > 0 || meta.postTokenBalances?.length > 0;
+
+	if (isSplTokenTransaction) {
+		return await getSplTransactionBalances(connection, meta);
+	}
+	return await getNativeTransactionBalances(transaction);
+};
+
+const getNativeTransactionBalances = async (transaction: ParsedTransaction) => {
+	// const _id = `${walletAddress}/token/${solana.solMint}`;
+	const token: TransactionHistory['token'] = {
+		network: Networks.solana,
+		metadata: solana.solMetadata,
+	};
+	const parsedInstruction = transaction.message.instructions.find(
+		(instruction) =>
+			'parsed' in instruction && instruction.parsed !== 'undefined',
+	) as ParsedInstruction;
+
+	const sender = (parsedInstruction?.parsed?.info?.source as string) || '';
+	const receiver =
+		(parsedInstruction?.parsed?.info?.destination as string) || '';
+	const amount =
+		((parsedInstruction?.parsed?.info?.lamports as number) || 0) / 10 ** 9;
+
+	return {
+		sender,
+		receiver,
+		amount,
+		token,
+	};
+};
+
+const getSplTransactionBalances = async (
+	connection: Connection,
+	meta: ParsedTransactionMeta,
+) => {
+	if (!meta.preTokenBalances || !meta.postTokenBalances) return;
+	if (meta.preTokenBalances.length === 0 || meta.postTokenBalances.length === 0)
+		return;
+
+	const { mint, sender, receiver, amount } = analyzeTokenBalances(
+		meta.preTokenBalances,
+		meta.postTokenBalances,
+	);
+
+	const token: TransactionHistory['token'] = {
+		network: Networks.solana,
+		metadata: await getMetadata(connection, mint),
+	};
+
+	return {
+		sender,
+		receiver,
+		amount: Math.abs(amount),
+		token,
+	};
+};
+
+const analyzeTokenBalances = (
+	preTokenBalances: TokenBalance[],
+	postTokenBalances: TokenBalance[],
+) => {
+	if (postTokenBalances.length > preTokenBalances.length) {
+		const referralPreBalance = preTokenBalances[0];
```

Comment:
Why does it include `referral` here

Reply:
this is my mistake, I mean referenced. Because `preTokenBalances` and `postTokenBalanced` are not always identical, I need to choose 1 `TokenBalance` as a referenced to perform calculation. Update to `referencedPreBalance` etc.